pulser-core/pulser/backend/operator.py

```diff
@@ -0,0 +1,140 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the abstract base class for a quantum operator."""
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Collection, Mapping, Sequence
+from typing import Generic, Type, TypeVar
+
+from pulser.backend.state import Eigenstate, State
+
+ArgScalarType = TypeVar("ArgScalarType")
+ReturnScalarType = TypeVar("ReturnScalarType")
+StateType = TypeVar("StateType", bound=State)
+OperatorType = TypeVar("OperatorType", bound="Operator")
+
+QuditOp = Mapping[str, ArgScalarType]  # single qudit operator
+TensorOp = Sequence[
+    tuple[QuditOp, Collection[int]]
+]  # QuditOp applied to set of qudits
+FullOp = Sequence[tuple[ArgScalarType, TensorOp]]  # weighted sum of TensorOp
+
+
+class Operator(ABC, Generic[ArgScalarType, ReturnScalarType, StateType]):
+    """Base class for a quantum operator."""
+
+    @abstractmethod
+    def apply_to(self, state: StateType, /) -> StateType:
+        """Apply the operator to a state.
+
+        Args:
+            state: The state to apply this operator to.
+
+        Returns:
+            The resulting state.
+        """
+        pass
+
+    @abstractmethod
+    def expect(self, state: StateType, /) -> ReturnScalarType:
+        """Compute the expectation value of self on the given state.
+
+        Args:
+            state: The state with which to compute.
+
+        Returns:
+            The expectation value.
+        """
+        pass
+
+    @abstractmethod
+    def __add__(self: OperatorType, other: OperatorType, /) -> OperatorType:
+        """Computes the sum of two operators.
+
+        Args:
+            other: The other operator.
+
+        Returns:
+            The summed operator.
+        """
+        pass
+
+    @abstractmethod
+    def __rmul__(self: OperatorType, scalar: ArgScalarType) -> OperatorType:
+        """Scale the operator by a scalar factor.
+
+        Args:
+            scalar: The scalar factor.
+
+        Returns:
+            The scaled operator.
+        """
+        pass
+
+    @abstractmethod
+    def __matmul__(self: OperatorType, other: OperatorType) -> OperatorType:
+        """Compose two operators where 'self' is applied after 'other'.
+
+        Args:
+            other: The operator to compose with self.
+
+        Returns:
+            The composed operator.
+        """
+        pass
+
+    @classmethod
+    @abstractmethod
+    def from_operator_repr(
+        cls: Type[OperatorType],
+        *,
+        eigenstates: Sequence[Eigenstate],
+        n_qudits: int,
+        operations: FullOp,
```

Comment:
in the notion page we had an extra argument here `operators: dict[str, QubitOp]` where you could define aliases for certain `QuditOp`s such as `{"X": {"rg": 1, "gr": 1}}`. I found this super useful and started depending on it in several places. Now I realize it's not in the method. Could we add it back in?

Reply:
Can't it be specfic to `MPO`? I'm not a big fan of it and chose not to have it in `QutipOperator` as I found it overcomplicated the implementation unnecessarily. 
I think that what you wanted could be achieved by defining an auxiliary variable instead, eg.
```
X = {"rg": 1, "gr": 1}
QutipOperator.from_operator_repr(
    eigenstates=("r", "g"),
    n_qudits=3,
    operations=[(1.0, [(X, {0, 2})]],
)
```

Reply:
That is fair.